pennylane/ops/op_math/condition.py

```diff
@@ -647,39 +665,134 @@ def wrapper(*args, **kwargs):
 ) -> None:
     """Ensure the collected abstract values match the expected ones."""
+    import jax  # pylint: disable=import-outside-toplevel
 
     if len(outvals) != len(expected_outvals):
-        raise ValueError(
+        msg = (
             f"Mismatch in number of output variables in {branch_type} branch"
-            f"{'' if index is None else ' #' + str(index)}: "
-            f"{len(outvals)} vs {len(expected_outvals)}"
+            f" #{branch_index}: {len(outvals)} vs {len(expected_outvals)} "
+            f" for {outvals} and {expected_outvals}"
         )
+        if jax.config.jax_dynamic_shapes:
+            msg += "\n This may be due to different sized shapes when dynamic shapes are enabled."
+        raise ValueError(msg)
 
     for i, (outval, expected_outval) in enumerate(zip(outvals, expected_outvals)):
-        if outval != expected_outval:
-            raise ValueError(
-                f"Mismatch in output abstract values in {branch_type} branch"
-                f"{'' if index is None else ' #' + str(index)} at position {i}: "
-                f"{outval} vs {expected_outval}"
-            )
+        if jax.config.jax_dynamic_shapes:
+            # we need to be a bit more manual with the comparison.
+            if type(outval) != type(expected_outval):  # pylint: disable=unidiomatic-typecheck
+                _shape_error(branch_type, branch_index, i, outval, expected_outval)
+            if getattr(outval, "dtype", None) != getattr(expected_outval, "dtype", None):
+                _shape_error(branch_type, branch_index, i, outval, expected_outval)
+
+            for s1, s2 in zip(getattr(outval, "shape", ()), getattr(expected_outval, "shape", ())):
+                if isinstance(s1, jax.core.Var) != isinstance(s2, jax.core.Var):
+                    _shape_error(branch_type, branch_index, i, outval, expected_outval)
+                elif isinstance(s1, int) and s1 != s2:
+                    _shape_error(branch_type, branch_index, i, outval, expected_outval)
+
+        elif outval != expected_outval:
+            _shape_error(branch_type, branch_index, i, outval, expected_outval)
+
+
+def _validate_jaxpr_returns(jaxpr_branches):
+    outvals_true = [out.aval for out in jaxpr_branches[0].outvars]
+    for idx, jaxpr_branch in enumerate(jaxpr_branches):
+
+        if idx == 0:
+            continue
+
+        if jaxpr_branch is None:
+            if outvals_true:
+                raise ValueError(
+                    "The false branch must be provided if the true branch returns any variables"
+                )
+            # this is tested, but coverage does not pick it up
+            continue  # pragma: no cover
+
+        outvals_branch = [out.aval for out in jaxpr_branch.outvars]
+        branch_type = "elif" if idx < len(jaxpr_branches) - 1 else "false"
+        _validate_abstract_values(outvals_branch, outvals_true, branch_type, idx - 1)

```

Comment:
Since we're validating avals, not concrete vals
```suggestion
    out_avals_true = [out.aval for out in jaxpr_branches[0].outvars]
    for idx, jaxpr_branch in enumerate(jaxpr_branches):

        if idx == 0:
            continue

        if jaxpr_branch is None:
            if out_avals_true:
                raise ValueError(
                    "The false branch must be provided if the true branch returns any variables"
                )
            # this is tested, but coverage does not pick it up
            continue  # pragma: no cover

        out_avals_branch = [out.aval for out in jaxpr_branch.outvars]
        branch_type = "elif" if idx < len(jaxpr_branches) - 1 else "false"
        _validate_abstract_values(out_avals_branch, out_avals_true, branch_type, idx - 1)

```